File: frontend/types/subscription.ts
export type BillingPeriod = "MONTHLY" | "QUARTERLY" | "YEARLY";

export type SubscriptionInvoiceStatus =
  | "PENDING_PAYMENT"
  | "PROOF_UPLOADED"
  | "PAID"
  | "REJECTED"
  | "CANCELLED";

export type PlanEntitlements = {
  max_products: number | null;
  max_product_images: number;
  product_video: boolean;
  custom_fields: boolean;
  discounts: boolean;
  analytics_max_days: number;
  guest_checkout: boolean;
  badge_trust: boolean;
  badge_premium: boolean;
  store_theme: boolean;
  excel_export: boolean;
  store_pages: boolean;
  priority_support: boolean;
  [key: string]: unknown;
};

export type SubscriptionPlan = {
  id: number;
  code: string;
  name_fa: string;
  monthly_price_toman: number;
  quarterly_price_toman: number;
  yearly_price_toman: number;
  is_recommended: boolean;
  sort_order: number;
  is_active: boolean;
  entitlements: PlanEntitlements;
};

export type SellerSubscription = {
  id: number;
  plan: SubscriptionPlan;
  status: "ACTIVE" | "PAST_DUE" | "CANCELLED";
  billing_period: BillingPeriod;
  current_period_start: string | null;
  current_period_end: string | null;
  started_at: string;
};

export type SellerSubscriptionDetail = {
  subscription: SellerSubscription;
  effective_plan: SubscriptionPlan;
  entitlements: PlanEntitlements;
  is_expired_fallback: boolean;
};

export type SubscriptionPaymentProof = {
  id: number;
  image_url: string;
  uploaded_at: string;
};

export type SubscriptionInvoice = {
  id: number;
  plan: SubscriptionPlan;
  period: BillingPeriod;
  amount_toman: number;
  status: SubscriptionInvoiceStatus;
  period_start: string | null;
  period_end: string | null;
  admin_note: string | null;
  created_at: string;
  proofs: SubscriptionPaymentProof[];
};

export type CardInstructions = {
  card_number: string;
  card_owner: string;
  card_bank: string;
  message: string;
};

export type SubscriptionCheckoutResponse = {
  invoice: SubscriptionInvoice;
  payment: {
    provider: string;
    mode: string;
    instructions: Record<string, unknown>;
    redirect_url: string | null;
    external_reference: string | null;
  };
  card_instructions: CardInstructions;
};

export type AdminSubscriptionInvoice = SubscriptionInvoice & {
  seller_user_id: number;
  seller_email: string | null;
  seller_full_name: string | null;
};

File: backend/app/schemas/subscription.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import BillingPeriod, SellerSubscriptionStatus, SubscriptionInvoiceStatus


class SubscriptionPlanResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    code: str
    name_fa: str
    monthly_price_toman: int
    quarterly_price_toman: int
    yearly_price_toman: int
    is_recommended: bool
    sort_order: int
    is_active: bool
    entitlements: dict[str, Any]


class SubscriptionPlanUpdate(BaseModel):
    name_fa: str | None = None
    monthly_price_toman: int | None = Field(default=None, ge=0)
    quarterly_price_toman: int | None = Field(default=None, ge=0)
    yearly_price_toman: int | None = Field(default=None, ge=0)
    is_recommended: bool | None = None
    sort_order: int | None = None
    is_active: bool | None = None
    entitlements: dict[str, Any] | None = None


class SellerSubscriptionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    plan: SubscriptionPlanResponse
    status: SellerSubscriptionStatus
    billing_period: BillingPeriod
    current_period_start: datetime | None
    current_period_end: datetime | None
    started_at: datetime


class SellerSubscriptionDetailResponse(BaseModel):
    subscription: SellerSubscriptionResponse
    effective_plan: SubscriptionPlanResponse
    entitlements: dict[str, Any]
    is_expired_fallback: bool


class SubscriptionCheckoutRequest(BaseModel):
    plan_code: str
    period: BillingPeriod = BillingPeriod.MONTHLY


class SubscriptionPaymentProofResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    image_url: str
    uploaded_at: datetime


class SubscriptionInvoiceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    plan: SubscriptionPlanResponse
    period: BillingPeriod
    amount_toman: int
    status: SubscriptionInvoiceStatus
    period_start: datetime | None
    period_end: datetime | None
    admin_note: str | None = None
    created_at: datetime
    proofs: list[SubscriptionPaymentProofResponse] = []


class CardInstructionsResponse(BaseModel):
    card_number: str
    card_owner: str
    card_bank: str
    message: str


class PaymentSessionResponse(BaseModel):
    provider: str
    mode: str
    instructions: dict[str, Any]
    redirect_url: str | None = None
    external_reference: str | None = None


class SubscriptionCheckoutResponse(BaseModel):
    invoice: SubscriptionInvoiceResponse
    payment: PaymentSessionResponse
    card_instructions: CardInstructionsResponse


class AdminSubscriptionInvoiceResponse(SubscriptionInvoiceResponse):
    seller_user_id: int
    seller_email: str | None = None
    seller_full_name: str | None = None


class AdminInvoiceActionRequest(BaseModel):
    admin_note: str | None = None


class AdminAssignPlanRequest(BaseModel):
    plan_code: str
    period: BillingPeriod = BillingPeriod.MONTHLY
    months: int = Field(default=1, ge=1, le=36)
